### ISSUE TITLE
DNS lookup on custom TCP-like stream?

### ISSUE THREAD

**Issue by Mygod** [bluejekyll/trust-dns#1100]
What is the best way to implement DNS lookup on a custom struct that implements `DnsStreamHandle`, e.g. wrapping a `tokio::net::UnixStream` or a proxy TCP stream, so that it can use all the features in this library, e.g. connection reuse?

**Comment by bluejekyll** [comment]
Which part of the library do you mean? The `trust-dns-resolver` library already does this, you can just configure the resolver to only use TCP/TLS/HTTPS, and then it would reuse those connections.

If you want to get a raw packet, then `trust-dns-client` is probably what you want, though that won't do anything like CNAME resolution, and you'll need to interpret the DNS Message to extract the data types you want.

**Comment by Mygod** [comment]
Ah there is a client library. I missed that.

Though my use case is slightly different. `TcpClientStream` is still restrictive for us, in particular, the `async fn` in `trust_dns_proto::tcp::Connect` only takes `SocketAddr`, however, we want the stream initialization to be done via some other process (`UnixStream` or proxy set up). We currently have re-implemented [a lot of stuff](https://github.com/shadowsocks/shadowsocks-rust/blob/1a219d32d22af7ab2db2a0dcddfa6497108612af/src/relay/dnsrelay/upstream.rs#L86) but it would be great if we can use the connection pooling in this library...

**Comment by bluejekyll** [comment]
I think I understand what you want, though I don't think the library is necessarily in the best state to make that easy. I think you could implement your own version of `TcpClientStream` and then you could use the https://docs.rs/trust-dns-client/0.19.5/trust_dns_client/client/struct.AsyncClient.html#method.connect for using it.

You may want to explore using https://docs.rs/trust-dns-proto/0.19.5/trust_dns_proto/xfer/dns_multiplexer/struct.DnsMultiplexer.html for multiplexing over a single TCP stream. Then there is the https://docs.rs/trust-dns-proto/0.19.5/trust_dns_proto/xfer/struct.DnsExchange.html which can be used to wrap a multiplexed stream and properly dispatch and convert responses back to the requesting task. While the Client itself is not used in the trust-dns-resolver, the exchange and multiplexer are.

**Comment by Mygod** [comment]
Thanks! By the way does this library support connection pooling?

**Comment by bluejekyll** [comment]
The Client doesn’t, but the Resolver does.

**Comment by Mygod** [comment]
Ah I have given up trying to integrate the client library. For now I think I will keep implementing our own client/server code along with `trust-dns-proto`...

**Comment by bluejekyll** [comment]
Btw, it is possible in the Resolver to use custom TCP streams, by way of implementing a custom `Runtime` and or `ConnectionProvider`

**Comment by bluejekyll** [comment]
Also, if you have notes on what was complicated, and what you would like to see changed in the Client, we should capture those, and maybe someone will have some energy to work on them.

**Comment by Mygod** [comment]
Yes I made a preliminary attempt [here](https://gist.github.com/Mygod/2532e73063986635ce1f43db4e5825dd). How do I integrate it with trust-dns so that I can use the connection pool etc?

**Comment by Jason5Lee** [comment]
I have a similar requirement where I want to resolve DNS through socks5 proxy.

**Comment by chengyuhui** [comment]
Most libraries dealing with streams accept something like `where S: AsyncRead + AsyncWrite + Unpin`, making them more flexible. I wonder if we could have something similar, like a `trait StreamFactory`?
